src/Symfony/Component/Form/Extension/Core/Type/MultiStepType.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form\Extension\Core\Type;
+
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+use Symfony\Component\OptionsResolver\Options;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @author Silas Joisten <silasjoisten@proton.me>
+ * @author Patrick Reimers <preimers@pm.me>
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+final class MultiStepType extends AbstractType
+{
+    public function configureOptions(OptionsResolver $resolver): void
+    {
+        $resolver
+            ->setRequired('steps')
+            ->setDefault('current_step', static function (Options $options): string {
+                /** @var array<string, mixed> $steps */
+                $steps = $options['steps'];
+                $firstStep = array_key_first($steps);
+                if (!\is_string($firstStep)) {
+                    throw new \InvalidArgumentException('The option "steps" must be an associative array.');
+                }
+
+                return $firstStep;
+            });
+    }
+
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        $currentStep = $options['steps'][$options['current_step']];
+
+        if (\is_callable($currentStep)) {
+            $currentStep($builder, $options);
+        } elseif (\is_string($currentStep)) {
+            if (!class_exists($currentStep)) {
+                throw new \InvalidArgumentException(\sprintf('The form class "%s" does not exist.', $currentStep));
+            }
+
+            if (!is_subclass_of($currentStep, AbstractType::class)) {
+                throw new \InvalidArgumentException(\sprintf('"%s" is not a form type.', $currentStep));
+            }
+
+            $builder->add($options['current_step'], $currentStep);
+        }
+    }
+
+    public function buildView(FormView $view, FormInterface $form, array $options): void
+    {
+        $view->vars['current_step'] = $options['current_step'];
+        $view->vars['steps'] = array_keys($options['steps']);
```

Comment:
we could compute the `current_step_number` from those other options, which will be helpful for custom rendering

Reply:
Steps are named-based in this PR afaik.. so I think we have everything already for custom rendering.

More generally, I'm don't think the step _numbers_ should be used here, at least nor "pre-supposed" nor required.

Everytime i had to do such dynamic funnel / quote forms / wizards... one big requirement was to compute the steps "live", depending on what was set before..

So if in step "personal information" a user fill "kids: yes"  they would see next a step regarding their kids... where another perso would jump the step. "step 3" has then no universal meaning, and does not allow to match a form type.

So i would not focus too much on step "indexes" or "count" here.

Reply:
I added some more helpers in the template. could be helpful especially in determining if a step is first or last. (This is just a readable property not writable) And this actually will make the code in the UX pr less.

Reply:
@silasjoisten thanks! I've another question, do you know how we can handle the dynamic definition of steps for the case @smnandre explained in their comment? For example, I want to dynamically add a new step based on the data entered in the current step's form

Reply:
Yes its not possible without having a storage. This is just the implementation of a very simple multi step form. 

However i think this formtype is flexible enough to implement this as a user of this form. The user has to decide for a storage anyway in order to use with plain php. Ig you have a storage, you can inject it to your type and construct the steps dynamically

Reply:
My question is. Do we want to provide a storage in form? Like session storage? Or is that userland stuff?

Reply:
I'd be fine IMO if we also introduced the form flow concept/utility to handle multi-step forms